pkg/k8shandler/collection.go

```diff
@@ -243,7 +265,75 @@ func isFluentdDaemonsetDifferent(current *apps.DaemonSet, desired *apps.DaemonSe
 
 	different := false
 
-	// TODO: populate this
+	if isImageDifference(current, desired) {
+		return desired, true
+	}
 
 	return current, different
 }
+
+func waitForDaemonSetReady(ds *apps.DaemonSet) error {
+
+	err := wait.Poll(retryInterval, timeout, func() (done bool, err error) {
+		err = sdk.Get(ds)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return false, fmt.Errorf("Failed to get Fluentd daemonset: %v", err)
+			}
+			return false, err
+		}
+
+		if int(ds.Status.DesiredNumberScheduled) == int(ds.Status.NumberReady) {
+			return true, nil
+		}
+
+		return false, nil
+	})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func isImageDifference(current *apps.DaemonSet, desired *apps.DaemonSet) bool {
+
+	for _, curr := range current.Spec.Template.Spec.Containers {
+		for _, des := range desired.Spec.Template.Spec.Containers {
+			// Only compare the images of containers with the same name
+			if curr.Name == des.Name {
+				if curr.Image != des.Image {
+					return true
+				}
+			}
+		}
+	}
+
+	return false
+}
+
+func isBufferFlushRequired(current *apps.DaemonSet, desired *apps.DaemonSet) bool {
+
+	currImage := strings.Split(current.Spec.Template.Spec.Containers[0].Image, ":")
+	desImage := strings.Split(desired.Spec.Template.Spec.Containers[0].Image, ":")
+
+	if len(currImage) != 2 || len(desImage) != 2 {
+		// we don't have versions here -- not sure how we would compare versions to determine
+		// need to flush buffers
+		return false
+	}
+
+	currVersion := currImage[1]
+	desVersion := desImage[1]
+
+	if strings.HasPrefix(currVersion, "v") {
+		currVersion = strings.Split(currVersion, "v")[1]
+	}
+
+	if strings.HasPrefix(desVersion, "v") {
+		desVersion = strings.Split(desVersion, "v")[1]
+	}
+
+	return (currVersion == "3.11" && desVersion == "4.0.0")
```

Comment:
@richm are we providing fluentd 1.x as our 4.0 image?

Reply:
yes